src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -33,55 +31,84 @@ internal abstract partial class AbstractNavigateToSearchService
+                {
+                    nameMatches.Clear();
+                    containerMatches.Clear();
 
                     cancellationToken.ThrowIfCancellationRequested();
-                    var declarationInfo = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
-
-                    foreach (var declaredSymbolInfo in declarationInfo.DeclaredSymbolInfos)
-                    {
-                        cancellationToken.ThrowIfCancellationRequested();
-                        var patternMatches = patternMatcher.GetMatches(
-                            declaredSymbolInfo.Name,
-                            declaredSymbolInfo.FullyQualifiedContainerName,
-                            includeMatchSpans: true);
-
-                        if (!patternMatches.IsEmpty)
-                        {
-                            result.Add(ConvertResult(containsDots, declaredSymbolInfo, document, patternMatches));
-                        }
+                    if (nameMatcher.AddMatches(declaredSymbolInfo.Name, nameMatches) &&
+                        containerMatcher?.AddMatches(declaredSymbolInfo.FullyQualifiedContainerName, containerMatches) != false)
+                    { 
+                        result.Add(ConvertResult(
+                            declaredSymbolInfo, document, nameMatches, containerMatches));
                     }
                 }
-
-                return result.ToImmutableAndFree();
             }
+
+            return result.ToImmutableAndFree();
         }
 
         private static INavigateToSearchResult ConvertResult(
-            bool containsDots, DeclaredSymbolInfo declaredSymbolInfo, 
-            Document document, PatternMatches matches)
+            DeclaredSymbolInfo declaredSymbolInfo, Document document,
+            ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches)
         {
-            var matchKind = GetNavigateToMatchKind(containsDots, matches);
+            var matchKind = GetNavigateToMatchKind(nameMatches);
 
             // A match is considered to be case sensitive if all its constituent pattern matches are
             // case sensitive. 
-            var isCaseSensitive = matches.All(m => m.IsCaseSensitive);
+            var isCaseSensitive = nameMatches.All(m => m.IsCaseSensitive) && containerMatches.All(m => m.IsCaseSensitive);
             var kind = GetItemKind(declaredSymbolInfo);
             var navigableItem = NavigableItemFactory.GetItemFromDeclaredSymbolInfo(declaredSymbolInfo, document);
 
             return new SearchResult(
-                document, declaredSymbolInfo, kind, matchKind, 
-                isCaseSensitive, navigableItem,
-                matches.CandidateMatches.SelectMany(m => m.MatchedSpans).ToImmutableArray());
+                document, declaredSymbolInfo, kind, matchKind, isCaseSensitive, navigableItem,
+                nameMatches.SelectMany(m => m.MatchedSpans).ToImmutableArray());
```

Comment:
💭 (You did not change this but) the reason for a `SelectMany` here is not obvious from reading this. Will need to dig deeper but a separate task for me.